src/event/vrl_target.rs

```diff
@@ -0,0 +1,398 @@
+use super::{Event, Metric, MetricKind, Value};
+use snafu::Snafu;
+use std::{collections::BTreeMap, convert::TryFrom, iter::FromIterator};
+use vrl::path::Segment;
+
+const VALID_METRIC_PATHS_SET: &str = ".name, .namespace, .timestamp, .kind, .tags";
+
+/// We can get the `type` of the metric in Remap, but can't set  it.
+const VALID_METRIC_PATHS_GET: &str = ".name, .namespace, .timestamp, .kind, .tags, .type";
+
+/// An adapter to turn `Event`s into `vr::Target`s
+#[derive(Debug, Clone)]
+pub enum VrlTarget {
+    LogEvent(vrl::Value),
+    Metric(Metric),
+}
+
+impl VrlTarget {
+    pub fn new(event: Event) -> Self {
+        match event {
+            // TODO optimize
```

Comment:
Do you have anything specific in mind here?

Reply:
I was able to take advantage of the existing `From` implementations in 3cf40a339 but I have a feeling this is doing more copying than we might like. I plan to run a benchmark or two to see if the overhead is noticeable. I'm curious to hear any suggestion though.

Reply:
Ah, I see. So the concern is having three steps of `.into` conversions instead of (if possible) one complete conversion?

Reply:
Well, it's more that the last `into` (`VrlTarget::LogEvent(value.into())`) ends up doing a `BTreeMap` clone to map the values. This appears to be where Vector ends up spending a bunch of time in my benches.

Reply:
I was able to resolve this overhead by, instead of storing a `vrl::Value` in my new `vrl::Target::Log` implementation, storing a `event::Value`. This avoids needing to do the btreemap copy to `vrl::Value` on instantiation and back to `event::Value` when converting back into a `LogEvent`.